{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate enrichment scores across tree - BEAST trees\n",
    "\n",
    "This is an initial attempt to break this code into reasonable chunks that can be coded, tested, and executed separately. In this notebook, we will be reading in a tree, enumerating all the mutations on that tree, and calculating enrichment scores for each of them. These enrichment scores are based on [this code](https://github.com/sheppardlab/pGWAS/blob/master/assomap_given_phylo.py), written for [this paper](https://www.nature.com/articles/s41467-018-07368-7#Sec10) detailed in lines 245-273 and calculated from the following contingency table as: \n",
    "\n",
    "|host|presence|absence|\n",
    "|:------|:-------|:------|\n",
    "|host 1|A|B| \n",
    "|host 2|C|D|\n",
    "\n",
    "where A, B, C, and D are counts of the mutation's presence and absence in host 1 and host 2. The odds ratio is then calculated as: `OR = (A * D)/(B * C)`\n",
    "\n",
    "In this notebook, this code is written for parsing a tree json format, output from Nextstrain. In subsequent notebooks, I will alter this for running on beast trees. \n",
    "\n",
    "\n",
    "### A NOTE ON BALTIC: \n",
    "Currently, in the posterior set of trees, mutations are annotated as traits as `&typeTrait=domestic,mutations=\"G730A,A846G,C1203A,A1278G\"`. However, baltic is not reading in all the mutations properly, instead only reading in the first mutation in the list. This is due to line 1116 of baltic, which attempts to find trait strings. In order to make this work, I added in a comma as an acceptable character in the 4th block of the string search. I saved the version with this small edit as `../baltic/baltic/baltic-modified-for-muts.py`. I will use that baltic version in this notebook. Additionally, for this to work, the typeTrait also needs to be a string. \n",
    "\n",
    "\n",
    "Also, a note on trees: as currently written, this does some switching between numNames and strain nameese. If the tree you are testing it on doesen't have those, the code will need to be changed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob, json\n",
    "import re,copy, imp\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "from io import StringIO\n",
    "\n",
    "# for this to work, you will need to download the most recent version of baltic, available here \n",
    "#bt = imp.load_source('baltic', '../baltic/baltic/baltic.py')\n",
    "bt = imp.load_source('baltic', '../baltic/baltic/baltic-modified-for-muts.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Infer each mutation that occurs across the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxa_lines(tree_path):    \n",
    "\n",
    "    lines_to_write = \"\"\n",
    "    with open(tree_path, 'rU') as infile:\n",
    "        for line in infile: ## iterate through each line\n",
    "            if 'state' not in line.lower(): #going to grab all the interesting stuff in the .trees file prior to the newick tree strings\n",
    "                lines_to_write = lines_to_write + line\n",
    "\n",
    "    return(lines_to_write)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_strain_to_number(taxa_lines):\n",
    "    \n",
    "    output_dict = {}\n",
    "    \n",
    "    translation_block = taxa_lines.split(\"Translate\\n\")[1]\n",
    "    translation_list = translation_block.replace(\"\\t\",\"\").split(\"\\n\")\n",
    "    \n",
    "    for t in translation_list: \n",
    "        information = t.lstrip().replace(\",\",\"\")  # remove leading white spaces and commas\n",
    "        \n",
    "        if len(information.split(\" \")) == 2:\n",
    "            numeric_id = information.split(\" \")[0]\n",
    "            strain_name = information.split(\" \")[1]\n",
    "        \n",
    "            output_dict[numeric_id] = strain_name\n",
    "            \n",
    "        else: \n",
    "            pass\n",
    "        \n",
    "    return(output_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_leaves_to_strains(input_leaves, strains_dict): \n",
    "    output_list = []\n",
    "    \n",
    "    for l in input_leaves: \n",
    "        strain_name = strains_dict[l]\n",
    "        output_list.append(strain_name)\n",
    "        \n",
    "    return(output_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_burnin_value(tree_path, burnin_percent):\n",
    "    with open(tree_path, 'rU') as infile:\n",
    "        numtrees = 0\n",
    "        for line in infile: ## iterate through each line\n",
    "            if 'state' in line.lower(): #going to grab all the interesting stuff in the .trees file prior to the newick tree strings\n",
    "                numtrees += 1\n",
    "    \n",
    "    burnin = numtrees * burnin_percent\n",
    "    return(burnin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"count the number of tips on the tree corresponding to each host category\"\"\"\n",
    "def return_all_host_tips(tree):\n",
    "    host_counts = {'human':0, 'domestic':0, 'wild':0}\n",
    "    \n",
    "    for k in tree.Objects: \n",
    "        if k.branchType == \"leaf\":\n",
    "            host = k.traits['typeTrait']\n",
    "            host_counts[host] += 1\n",
    "    return(host_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"given a branch and gene, return the mutations present on that branch\"\"\"\n",
    "\n",
    "def return_muts_on_branch(branch):\n",
    "    muts = []\n",
    "    \n",
    "    if 'mutations' in branch.traits:\n",
    "        muts = branch.traits['mutations'].split(\",\")\n",
    "                            \n",
    "    return(muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"this function does 2 things: 1. for each branch, it records the branch name and its branch length in a \n",
    "dictionary; 2. it adds up the total branch length on the tree. For the beast trees, we only have branch lengths \n",
    "in time. However, we can get a reasonable branch length (and I think this is perfectly reasonabble for this purpose)\n",
    "by just summing the total mutations on the branch and dividing by the total number of sites. The only purpose \n",
    "in this analysis for the total tree branch length is to get an idea of the number of mutations that should \n",
    "occur across the tree. So this should work.\"\"\"\n",
    "\n",
    "def return_total_tree_branch_length(tree, n_sites_alignment):\n",
    "    total_branch_length = 0\n",
    "    branch_lengths = {}\n",
    "    \n",
    "    for k in tree.Objects:\n",
    "        muts_on_branch = return_muts_on_branch(k)\n",
    "        branch_length_time = k.length\n",
    "        branch_length_divergence = len(muts_on_branch)/n_sites_alignment\n",
    "                \n",
    "        total_branch_length += branch_length_divergence\n",
    "        branch_lengths[k] = branch_length_divergence\n",
    "    \n",
    "    return(total_branch_length, branch_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Traverse the tree from root to tip. On each branch, tips and nodes, gather every nucleotide and amino acid \n",
    "mutation\"\"\"\n",
    "\n",
    "def gather_all_mut_on_tree(tree):\n",
    "    \n",
    "    all_nt_muts = []\n",
    "    \n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits:\n",
    "            nt_muts = k.traits['mutations'].split(\",\")\n",
    "            all_nt_muts.extend(nt_muts)\n",
    "                \n",
    "    all_nt_muts = list(set(all_nt_muts))\n",
    "    \n",
    "    return(all_nt_muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"return the total number of times that the mutation arises on the phylogeny. This includes instances of mutation \n",
    "on internal nodes and on tips and counts each with the same weight\"\"\"\n",
    "\n",
    "def return_number_times_on_tree(tree, mut):\n",
    "    times_on_tree = 0\n",
    "    \n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits:\n",
    "        \n",
    "            nt_muts = k.traits['mutations'].split(\",\")\n",
    "            if mut in nt_muts: \n",
    "                times_on_tree += 1\n",
    "                                                \n",
    "    return(times_on_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"return the total number of times that the mutation arises on the phylogeny. This includes instances of mutation \n",
    "on internal nodes and on tips and counts each with the same weight\"\"\"\n",
    "\n",
    "def return_branch_length_mut_on_tree(tree, mut,n_sites_alignment):\n",
    "    branch_length = 0\n",
    "            \n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits:\n",
    "            nt_muts = k.traits['mutations'].split(\",\")\n",
    "                \n",
    "            # if the mutation arises on this branch\n",
    "            if mut in nt_muts: \n",
    "                branch_length_div = len(nt_muts)/n_sites_alignment\n",
    "                    \n",
    "                branch_length += branch_length_div\n",
    "                    \n",
    "    return(branch_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Given a starting internal node, and a tip you would like to end at, traverse the full path from that node to\n",
    "tip. Along the way, gather mutations that occur along that path. Once you have reached the ending \n",
    "tip, return the list of mutations that fell along that path. Input for the ending tip here is a tip name, while \n",
    "the starting node is a node object\"\"\"\n",
    "\n",
    "def return_all_muts_on_path_to_tip(starting_node, ending_tip, muts, strains_dict):\n",
    "    \n",
    "    # set an empty list of mutations and enumerate the children of the starting node; children can be tips or nodes\n",
    "    children = starting_node.children\n",
    "    \n",
    "    for child in children:\n",
    "        local_muts = []\n",
    "        \n",
    "        \"\"\"if the child is a leaf: if leaf is the target end tip, add the mutations that occur on that branch to \n",
    "        the list and return the list; if leaf is not the target end tip, move on\"\"\"\n",
    "        \"\"\"if the child is an internal node: first, test whether that child node contains the target tips in its \n",
    "        children. child.leaves will output a list of the names of all tips descending from that node. If not, pass. \n",
    "        if the node does contain the target end tip in its leaves, keep traversing down that node recursively, \n",
    "        collecting mutations as you go\"\"\"\n",
    "\n",
    "        if child.branchType == \"leaf\":\n",
    "            if child.name != ending_tip:\n",
    "                pass\n",
    "            elif child.name == ending_tip:\n",
    "                host = child.traits[\"typeTrait\"]\n",
    "                local_muts = return_muts_on_branch(child)\n",
    "                muts.extend(local_muts)\n",
    "                return(host, muts)\n",
    "        \n",
    "        elif child.branchType == \"node\":\n",
    "            strain_leaves = convert_leaves_to_strains(child.leaves, strains_dict)\n",
    "            if ending_tip not in strain_leaves:\n",
    "                pass\n",
    "            else:\n",
    "                local_muts = return_muts_on_branch(child)\n",
    "                muts.extend(local_muts)\n",
    "                host, muts = return_all_muts_on_path_to_tip(child, ending_tip, muts, strains_dict)\n",
    "    \n",
    "    return(host, muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"at times, will need to check whether the revertant mutation occcurs downstream. Return the revertant mutation\"\"\"\n",
    "\n",
    "def return_opposite_mutation(mut):\n",
    "    \n",
    "    site = mut[1:-1]\n",
    "    ref = mut[0]\n",
    "    alt = mut[-1]\n",
    "    opposite = alt+site+ref\n",
    "    \n",
    "    return(opposite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"given a tree, mutation, and gene, return the number of times that mutation is present in each host\"\"\"\n",
    "\n",
    "def return_host_distribution_mutation(tree, mut, strains_dict):\n",
    "    \n",
    "    host_counts_dict = {'human':0, 'domestic':0, 'wild':0}\n",
    "    back_mutation = return_opposite_mutation(mut)\n",
    "    \n",
    "    # iterate through tree\n",
    "    for k in tree.Objects:\n",
    "        if 'mutations' in k.traits:\n",
    "            nt_muts = k.traits['mutations'].split(\",\")\n",
    "                \n",
    "            # if we have reached a node or tip in the tree with the target mutation, enumerate descendants\n",
    "            if mut in nt_muts: \n",
    "                    \n",
    "                # if the mutation occurs on a leaf, record the host and move on \n",
    "                if k.branchType == 'leaf':\n",
    "                    host = k.traits['typeTrait']\n",
    "                    host_counts_dict[host] += 1\n",
    "                    \n",
    "                # else, if the mutation occurs on a node, traverse the children and return host\n",
    "                elif k.branchType == \"node\":\n",
    "                    all_leaves = k.leaves\n",
    "                    for leaf in all_leaves: \n",
    "                        muts = []\n",
    "                        strain_name = strains_dict[leaf]\n",
    "                        host, muts = return_all_muts_on_path_to_tip(k, strain_name, muts, strains_dict)\n",
    "                        \n",
    "                        if back_mutation in muts: \n",
    "                            pass\n",
    "                        elif back_mutation and mut in muts:  # if both the mutation and backmutation occur, print\n",
    "                            print(\"something odd happened\",leaf, back_mutation, mut)\n",
    "                        else:\n",
    "                            host_counts_dict[host] += 1\n",
    "                                \n",
    "    return(host_counts_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate the enrichment scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"calculate an enrichment score for an individual mutation, based on the counts across hosts\"\"\"\n",
    "\n",
    "def calculate_enrichment_score_counts(mut_counts_dict, host1, host2, host_counts):\n",
    "    total_host1_tree = host_counts[host1]\n",
    "    total_host2_tree = host_counts[host2]\n",
    "    \n",
    "    mut_host1 = mut_counts_dict[host1]\n",
    "    mut_host2 = mut_counts_dict[host2]\n",
    "    \n",
    "    # this is calculating this table as counts\n",
    "    presence_host1 = mut_host1\n",
    "    absence_host1 = total_host1_tree - mut_host1\n",
    "    presence_host2 = mut_host2\n",
    "    absence_host2 = total_host2_tree - mut_host2\n",
    "    \n",
    "    if presence_host2 == 0:\n",
    "        presence_host2 = 1\n",
    "    if absence_host1 == 0:\n",
    "        absence_host1 = 1\n",
    "\n",
    "    # this score is calculated in terms of its enrichment in host 1\n",
    "    score = (presence_host1 * absence_host2)/(presence_host2 * absence_host1)\n",
    "#     score = (presence_host1 + absence_host2) - (presence_host2 + absence_host1)\n",
    "    return(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"calculate an enrichment score for an individual mutation, based on the counts across hosts\"\"\"\n",
    "\n",
    "def calculate_enrichment_score_proportions(mut_counts_dict, host1, host2, host_counts):\n",
    "    total_host1_tree = host_counts[host1]\n",
    "    total_host2_tree = host_counts[host2]\n",
    "    \n",
    "    mut_host1 = mut_counts_dict[host1]\n",
    "    mut_host2 = mut_counts_dict[host2]\n",
    "    \n",
    "    total_tips_in_tree = total_host1_tree + total_host2_tree\n",
    "    \n",
    "    # this is calculating this table as proportions\n",
    "    presence_host1 = (mut_host1)/total_tips_in_tree\n",
    "    absence_host1 = (total_host1_tree - mut_host1)/total_tips_in_tree\n",
    "    presence_host2 = (mut_host2)/total_tips_in_tree\n",
    "    absence_host2 = (total_host2_tree - mut_host2)/total_tips_in_tree\n",
    "    \n",
    "#     if presence_host2 == 0:\n",
    "#         presence_host2 = 1\n",
    "#     if absence_host1 == 0:\n",
    "#         absence_host1 = 1\n",
    "\n",
    "    # this score is calculated in terms of its enrichment in host 1\n",
    "    #score = (presence_host1 * absence_host2)/(presence_host2 * absence_host1)\n",
    "    score = (presence_host1 + absence_host2) - (presence_host2 + absence_host1)\n",
    "    return(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"for a tree and all amino acid mutations, calculate the enrichment scores across the tree\"\"\"\n",
    "def calculate_enrichment_scores(tree, nt_muts, host1, host2, min_required_count, method, host_counts, n_sites_alignment, strains_dict):\n",
    "    scores = []\n",
    "    scores_dict = {}\n",
    "    times_detected_dict = {}\n",
    "    branch_lengths_dict = {}\n",
    "    host_counts_dict2 = {}\n",
    "    \n",
    "    if method == \"counts\":\n",
    "        enrichment_calculation_function = calculate_enrichment_score_counts\n",
    "    elif method == \"proportions\":\n",
    "        enrichment_calculation_function = calculate_enrichment_score_proportions\n",
    "    \n",
    "    for n in nt_muts:\n",
    "        times_detected = return_number_times_on_tree(tree, n)\n",
    "        times_detected_dict[n] = times_detected\n",
    "\n",
    "        branch_length_mut = return_branch_length_mut_on_tree(tree, n, n_sites_alignment)\n",
    "        branch_lengths_dict[n] = branch_length_mut\n",
    "\n",
    "        host_counts_dict = return_host_distribution_mutation(tree, n, strains_dict)\n",
    "        host_counts_dict2[n] = host_counts_dict\n",
    "        total_tips_with_mut = host_counts_dict['human'] + host_counts_dict['domestic'] + host_counts_dict['wild']\n",
    "\n",
    "        if total_tips_with_mut > min_required_count:\n",
    "            enrichment_score = enrichment_calculation_function(host_counts_dict, host1,host2, host_counts)\n",
    "            #print(enrichment_score, total_tips_with_mut, host_counts_dict)\n",
    "            scores.append(enrichment_score)\n",
    "            scores_dict[n] = enrichment_score\n",
    "            \n",
    "            \n",
    "    return(scores, scores_dict, times_detected_dict, branch_lengths_dict, host_counts_dict2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_on_posterior_trees(all_trees, burnin):\n",
    "    start_time = time.time()\n",
    "\n",
    "    with open(all_trees, \"r\") as infile:\n",
    "\n",
    "        tree_counter = 0\n",
    "\n",
    "        for line in infile:\n",
    "            if 'tree STATE_' in line:\n",
    "                \n",
    "                if tree_counter < 2:\n",
    "                #if tree_counter >= burnin:\n",
    "                    tree_counter += 1\n",
    "                    temp_tree = StringIO(taxa_lines + line)\n",
    "                    tree = bt.loadNexus(temp_tree)\n",
    "\n",
    "    # print the amount of time this took\n",
    "    total_time_seconds = time.time() - start_time\n",
    "    total_time_minutes = total_time_seconds/60\n",
    "    print(\"this took\", total_time_seconds, \"seconds (\", total_time_minutes,\" minutes) to run on\", tree_counter, \"trees\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/envs/LHM-basics/lib/python3.7/site-packages/ipykernel_launcher.py:4: DeprecationWarning: 'U' mode is deprecated\n",
      "  after removing the cwd from sys.path.\n",
      "/Users/lmoncla/anaconda/envs/LHM-basics/lib/python3.7/site-packages/ipykernel_launcher.py:2: DeprecationWarning: 'U' mode is deprecated\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "all_trees = \"../test-data/2021-03-15-mascot-3deme-no-human-mig-TEST.combined.trees\"\n",
    "burnin_percent = 0\n",
    "\n",
    "taxa_lines = get_taxa_lines(all_trees)\n",
    "strains_dict = convert_strain_to_number(taxa_lines)\n",
    "\n",
    "burnin = get_burnin_value(all_trees, burnin_percent)\n",
    "n_sites_alignment = 1762\n",
    "\n",
    "with open(all_trees, \"r\") as infile:\n",
    "    muts = []\n",
    "    tree_counter = 0\n",
    "\n",
    "    for line in infile:\n",
    "        if 'tree STATE_' in line:\n",
    "                \n",
    "            if tree_counter >= burnin:\n",
    "                tree_counter += 1\n",
    "                print(tree_counter)\n",
    "                temp_tree = StringIO(taxa_lines + line)\n",
    "                tree = bt.loadNexus(temp_tree)      \n",
    "                        \n",
    "                host_counts = return_all_host_tips(tree)\n",
    "                x,y = return_total_tree_branch_length(tree, n_sites_alignment)\n",
    "                all_nt_muts = gather_all_mut_on_tree(tree)\n",
    "                times_on_tree = return_number_times_on_tree(tree, mut)\n",
    "                branch_length_mut = return_branch_length_mut_on_tree(tree, mut,n_sites_alignment)\n",
    "                xdict = return_host_distribution_mutation(tree, mut, strains_dict)\n",
    "    \n",
    "                scores, scores_dict, times_detected_dict, branch_lengths_dict, host_counts_dict2 = calculate_enrichment_scores(tree, all_nt_muts, \"human\",\"domestic\", 1, \"counts\",host_counts, n_sites_alignment, strains_dict)\n",
    "\n",
    "scores, scores_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LHM-basics (python3)",
   "language": "python",
   "name": "lhm-basics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
